src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java

```diff
@@ -1,64 +1,216 @@
+                    extImpTtx.getProductId(), extImpTtx.getZoneId(), errors));
+        } catch (PreBidException e) {
+            errors.add(BidderError.badInput(e.getMessage()));
+        }
+
+        final BidRequest modifiedRequest = requestToUpdate.build();
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers())
+                .payload(modifiedRequest)
+                .body(mapper.encode(modifiedRequest))
                 .build();
     }
 
+    private ExtImpTtx parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), TTX_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private List<Imp> updateImpList(List<Imp> requestImps, Imp firstImp, String productId,
+                                    String zoneId, List<BidderError> errors) {
+        final Imp.ImpBuilder modifiedFirstImp = firstImp.toBuilder();
+        modifiedFirstImp.ext(createImpExt(productId, zoneId)).build();
+
+        final Video video = firstImp.getVideo();
+        try {
+            if (video != null) {
+                modifiedFirstImp.video(updateVideo(video, productId));
+            }
+        } catch (PreBidException e) {
+            errors.add(BidderError.badInput(e.getMessage()));
+        }
+
+        final List<Imp> imps = new ArrayList<>(requestImps);
+        imps.set(0, modifiedFirstImp.build());
+        return imps;
+    }
+
+    private ObjectNode createImpExt(String productId, String zoneId) {
+        final TtxImpExt ttxImpExt = TtxImpExt.of(TtxImpExtTtx.of(productId, StringUtils.stripToNull(zoneId)));
+        return mapper.mapper().valueToTree(ttxImpExt);
+    }
+
+    private Video updateVideo(Video video, String productId) {
+        if (isZeroOrNullInteger(video.getW())
+                || isZeroOrNullInteger(video.getH())
+                || CollectionUtils.isEmpty(video.getProtocols())
+                || CollectionUtils.isEmpty(video.getMimes())
+                || CollectionUtils.isEmpty(video.getPlaybackmethod())
+        ) {
+            throw new PreBidException("One or more invalid or missing video field(s) "
+                    + "w, h, protocols, mimes, playbackmethod");
+        }
+
+        return modifyVideo(video, productId);
+    }
+
+    private Video modifyVideo(Video video, String productId) {
+        final Integer resolvedPlacement = resolvePlacement(video.getPlacement(), productId);
+        final Integer resolvedStartDelay = resolveStartDelay(productId);
+
+        return (resolvedPlacement != null || resolvedStartDelay != null)
+                ? video.toBuilder()
+                .startdelay(resolvedStartDelay != null ? resolvedStartDelay : video.getStartdelay())
+                .placement(resolvedPlacement).build()
```

Comment:
Should the original `placement` be used if null resolved?

Reply:
Idea is that this block will never be called if resolvedPlacement is null, but it's only for now. Thanks for catching this))